### ISSUE TITLE
How to deal with failed rollouts?

### ISSUE THREAD

**Issue by dalek-who** [microsoft/agent-lightning#448]
If a rollout fails due to system issues (such as network connection failure), should I return `None` in Agent.rollout? Can verl automatically skip such failed rollouts?

Additionally, if there are too many failed rollouts in a batch (e.g., due to temporary network instability), can I skip the entire batch directly? Are there any configurations for this?

**Comment by ultmaster** [comment]
In the rollout there is a RolloutConfig which you can configure the retry and max attempts.

If your experiment contains a lot of network failures, I guess you might want to fix your network first. Simply retrying won't automatically repair that kind of issues.

**Comment by dalek-who** [comment]
Currently, max_attempts for RolloutConfig is configured in `agentlightning/verl/daemon.py/AgentLightningTrainer._async_set_up`. It would be great if the RolloutConfig could be passed directly to the Trainer. This would make customizing rollouts much more straightforward.

**Comment by dalek-who** [comment]
My original question was: If an exception (e.g., due to a network error) occurs during a rollout, how can I drop this rollout entirely without retrying?
Currently, the default behavior in AgentLightning is: For an agent composed of multiple sub-agents, if the k-th sub-agent throws an exception, the prompts and responses from the previous k-1 sub-agents are still recorded and will be used for training (if there is no reward, AgentLightning will default the reward to 0. This may lead to unexpected behavior in cases where the minimum reward can be less than 0).
However, I have now found a way to completely skip the rollout that contains an exception. This method requires the use of `emit_exception` and `adapter`. The principle is: Even though the spans of the first k-1 sub-agents are collected by default, as long as an empty Triplet is returned, the entire rollout can be skipped.

The adapter:
```python
from typing import Any, Dict, Optional, List, Sequence, Union, Literal

import agentlightning as agl
from agentlightning.types import Attributes, SpanLike
from opentelemetry.semconv.attributes import exception_attributes
from agentlightning.semconv import AGL_EXCEPTION
from opentelemetry.sdk.trace import ReadableSpan
from agentlightning.types import Span, Triplet


def extract_exceptions_from_spans(spans: Sequence[SpanLike]) -> List[Dict[str, Any]]:
    exceptions = []
    for span in spans:
        if span.name == AGL_EXCEPTION:
            attrs = span.attributes or {}
            exceptions.append(attrs)
    return exceptions


class MyTracerTraceToTriplet(agl.TracerTraceToTriplet):

    def __init__(self, drop_exception_rollout: bool, *args: Any, **kwargs: Any):
        super().__init__(*args, **kwargs)
        self.drop_exception_rollout = drop_exception_rollout

    def adapt(self, source: Union[Sequence[Span], Sequence[ReadableSpan]], /) -> List[Triplet]:  # type: ignore
        exceptions = extract_exceptions_from_spans(spans=source)
        if exceptions and self.drop_exception_rollout:
            triplets = []
        else:
            triplets = super().adapt(source)

        return triplets
```

Agent rollout:
```python
# code for invoke llm
try:
    result = llm.invoke(prompt)
    return result

except Exception as e:
    agl.emit_exception(e, attributes={"training": state["training"], "stage_name": stage_name})
    raise
```

Trainer:
```python
adapter = MyTracerTraceToTriplet(
    drop_exception_rollout=True,
    ...  # other config
)
trainer = agl.Trainer(
    adapter=adapter,
    ...
)    
```

**Comment by ultmaster** [comment]
I think that's a valid approach for handling such issue.

Agent-lightning is not (at least currently) opinionated towards different methods of handling failures. You can customize algorithms, customized adapters. Anything you can think of should not be restricted by the framework -- that's our goal.

As for which method is the best practice, that needs more experiments and feedbacks from community.

**Comment by AdithyaKotian** [comment]
From looking at the current code:

- Rollout retries are handled through `RolloutConfig` (for example `max_attempts`, retry conditions, and timeouts).
- If a rollout fails due to a system issue like a network error, returning `None` is fine, but it still counts as an attempt and follows the retry limits.
- Failures are handled at the individual rollout level.There doesn’t seem to be a built in way to automatically skip an entire batch when many rollouts fail.
- If failures keep happening (e.g. frequent network issues), it’s likely an infrastructure problem rather than something retries are meant to hide.

It might be helpful to document some best practices around when to retry versus when to stop and fix the underlying issue.

**Comment by ultmaster** [comment]
@AdithyaKotian That's a good summary. Maybe you can help document it somewhere?
